File: Go/Lesson_107_implement_queue_using_stacks.go
//go:build ignore

// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 107 -- Implement Queue using Stacks
// Category   : Stack
// Difficulty : Easy
// Study Plan : Day 54
// =============================================================
//
// QUESTION:
//   Implement a FIFO queue using only two stacks.
// =============================================================

package main
import "fmt"
type MyQueue struct { a, b []int }
func (q *MyQueue) Push(x int){ q.a = append(q.a, x) }
func (q *MyQueue) Pop() int { q.Peek(); n := len(q.b); x := q.b[n-1]; q.b = q.b[:n-1]; return x }
func (q *MyQueue) Peek() int {
    if len(q.b)==0 { for len(q.a)>0 { n:=len(q.a); q.b=append(q.b, q.a[n-1]); q.a=q.a[:n-1] } }
    return q.b[len(q.b)-1]
}
func (q *MyQueue) Empty() bool { return len(q.a)==0 && len(q.b)==0 }
func main(){ q := &MyQueue{}; q.Push(1); q.Push(2);
    fmt.Println(q.Peek(), q.Pop(), q.Empty()) }

File: Python/Lesson_107_implement_queue_using_stacks.py
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 107 -- Implement Queue using Stacks
# Category   : Stack
# Difficulty : Easy
# Study Plan : Day 54
# =============================================================
#
# QUESTION:
#   Implement a FIFO queue using only two stacks.
# =============================================================

class MyQueue:
    def __init__(self): self.a=[]; self.b=[]
    def push(self, x): self.a.append(x)
    def pop(self):
        if not self.b:
            while self.a: self.b.append(self.a.pop())
        return self.b.pop()
    def peek(self):
        if not self.b:
            while self.a: self.b.append(self.a.pop())
        return self.b[-1]
    def empty(self): return not self.a and not self.b

if __name__ == "__main__":
    q=MyQueue(); q.push(1); q.push(2); print(q.peek()); print(q.pop()); print(q.empty())
